docs/struct.md

```diff
@@ -2,9 +2,170 @@
+│ {1.0,3.0,16744447}     │
+│ {1.25,2.75,4286513407} │
+│ {1.5,2.5,2139062271}   │
+│ {1.75,2.25,3638089983} │
+└────────────────────────┘
+```
+
+Now's an excellent time to ask: how's that stored in memory? Before we get to that answer, consider this other scenario, in Rust:
+
+```rust
+let v: [Point2D; 4] = [
+    Point2D {
+        x: 1.0,
+        y: 3.0,
+        rgb: 0x00FF7FFFu32,
+    },
+    Point2D {
+        x: 1.25,
+        y: 2.75,
+        rgb: 0xFF7F00FFu32,
+    },
+    Point2D {
+        x: 1.5,
+        y: 2.5,
+        rgb: 0x7F7F7FFFu32,
+    },
+    Point2D {
+        x: 1.75,
+        y: 2.25,
+        rgb: 0xD8D8D8FFu32,
+    },
+];
+```
+
+How's this one stored in memory? You might find that answer easier, it's an array of struct instances, so we have the `x`, `y` and `rgba` fields contiguously in memory, like that:
+
+![Diagram showing the contiguous layout memory of an array of structs in Rust](assets/struct_array_memory_layout.png){ style="display: block; margin: 0 auto;" }
+
+This is consistent with how C, C++, and many other languages store structs in memory.
+How's our struct-in-a-DataFrame different?
+
+Polars follows the Arrow protocol for structs, which means each field of the struct is stored in a Series, backed by chunks. Each chunk is contiguous in memory.
+That makes it so, for instance, assuming a single chunk, the address after 64 bits from the start of the first x field holds the x value for the next row, not the y field of the same row.
+It's also important to mention the chunks composing the Series are stored on the heap. In a scenario in which we have a single chunk for each field, this is how things would look like:
+
+![Diagram showing how Series backing struct fields are stored in memory](assets/structchunked_fields_memory_layout.png){ style="display: block; margin: 0 auto;" }
+
+Since we never modified the DataFrame after creating it, it is the case of our initial example, no more chunks were allocated.
+
+---
+
+Now that we have a better idea of how things work under the hood, let's jump to a practical plugin that takes a struct, and just prints the Series corresponding to each field - it'll return the same struct passed as input, with no alteration.
+
+First things first - this time we're gonna see somthing new. Polars does not allow us to write:
+
+```rust
+#[polars_expr(output_type=Struct)]
+fn print_struct_fields(inputs: &[Series]) -> PolarsResult<Series> { ... }
+```
+
+The way we inform a struct Series is being returned is a bit cumbersome - we do so by defining a separate function:
+
+```rust
+fn struct_point_2d_output(input_fields: &[Field]) -> PolarsResult<Field> {
+    let field = &input_fields[0];
+    match field.dtype() {
+        DataType::Struct(fields) => {
+            Ok(Field::new("struct_point_2d".into(), DataType::Struct(fields.clone())))
+        }
+        // Assuming a controlled environment in which we know the input
+        _ => unreachable!(),
```

Comment:
not sure we should teach this, else someone risks copy-and-pasting it into their code 😄 can we use the `polars_bail` macro?